src/org-folding-and-outline-provider.ts

```diff
@@ -0,0 +1,137 @@
+
+    constructor() {
+        this.documentStateRegistry = new WeakMap();
+    }
+
+    provideFoldingRanges(document: TextDocument, token: CancellationToken): ProviderResult<FoldingRange[]> {
+        let state = this.getOrCreateDocumentState(document);
+        return state.getRanges(document);
+    }
+
+    provideDocumentSymbols(document: TextDocument, token: CancellationToken): ProviderResult<SymbolInformation[]> {
+        let state = this.getOrCreateDocumentState(document);
+        return state.getSymbols(document);
+    }
+
+    private getOrCreateDocumentState(document: TextDocument): OrgFoldingAndOutlineDocumentState {
+        let state = this.documentStateRegistry.get(document);
+        if (!state) {
+            state = new OrgFoldingAndOutlineDocumentState();
+            this.documentStateRegistry.set(document, state);
+        }
+        return state;
+    }
+}
+
+class OrgFoldingAndOutlineDocumentState {
+    private computedForDocumentVersion: number = null;
+    private ranges: FoldingRange[] = [];
+    private symbols: SymbolInformation[] = [];
+
+    getRanges(document: TextDocument): FoldingRange[] {
+        this.compute(document);
+        return this.ranges;
+    }
+
+    getSymbols(document: TextDocument): SymbolInformation[] {
+        this.compute(document);
+        return this.symbols;
+    }
+
+    private compute(document: TextDocument) {
+        if (document.version === this.computedForDocumentVersion) {
+            return;
+        }
+        this.computedForDocumentVersion = document.version;
+        this.ranges = [];
+        this.symbols = [];
+
+        const count = document.lineCount;
+        let stack: Chunk[] = [];
+        let inBlock = false;
+
+        for (let lineNumber = 0; lineNumber < count; lineNumber++) {
+            const element = document.lineAt(lineNumber);
+            const text = element.text;
+
+            if (inBlock) {
+                if (isBlockEndLine(text)) {
+                    inBlock = false;
+                    if (stack.length > 0 && stack[stack.length - 1].type === ChunkType.BLOCK) {
+                        const top = stack.pop();
+                        this.createSection(top, lineNumber);
+                    }
+                }
+            } else if (isBlockStartLine(text)) {
+                inBlock = true;
+                let title = this.extractBlockTitle(text);
+                stack.push({ type: ChunkType.BLOCK, title, level: Number.MAX_SAFE_INTEGER, startLine: lineNumber });
+            } else if (isHeaderLine(text)) {
+                let currentLevel = getStarPrefixCount(text);
+
+                // close previous sections
+                while (stack.length > 0 && stack[stack.length - 1].level >= currentLevel) {
+                    const top = stack.pop();
+                    this.createSection(top, lineNumber - 1);
+                }
+
+                let title = text.substr(text.indexOf(' ') + 1);
```

Comment:
const too

Reply:
In a future version we could extract a `getTitle` function that would take care of the TODOs and trailing `:tags:`